flash/vision/style_transfer/data.py

```diff
@@ -0,0 +1,111 @@
+import functools
+import pathlib
+from typing import Any, Callable, Dict, Optional, Tuple, Union
+
+import torchvision
+from torch import nn
+from torchvision import transforms
+
+from flash.data.data_source import DefaultDataKeys, DefaultDataSources
+from flash.data.process import Preprocess
+from flash.data.transforms import ApplyToKeys
+from flash.vision.classification import ImageClassificationData, ImageClassificationPreprocess
+
+from ._utils import raise_not_supported
+
+__all__ = ["StyleTransferPreprocess", "StyleTransferData"]
+
+
+def _apply_to_input(default_transforms_fn) -> Callable[..., Dict[str, ApplyToKeys]]:
+
+    @functools.wraps(default_transforms_fn)
+    def wrapper(*args: Any, **kwargs: Any) -> Dict[str, ApplyToKeys]:
+        default_transforms = default_transforms_fn(*args, **kwargs)
+        return {hook: ApplyToKeys(DefaultDataKeys.INPUT, transform) for hook, transform in default_transforms.items()}
+
+    return wrapper
+
+
+class StyleTransferPreprocess(ImageClassificationPreprocess):
+
+    def __init__(
+        self,
+        train_transform: Optional[Union[Dict[str, Callable]]] = None,
+        val_transform: Optional[Union[Dict[str, Callable]]] = None,
+        test_transform: Optional[Union[Dict[str, Callable]]] = None,
+        predict_transform: Optional[Union[Dict[str, Callable]]] = None,
+        image_size: Union[int, Tuple[int, int]] = 256,
+    ):
+        if val_transform:
+            raise_not_supported("validation")
+        if test_transform:
+            raise_not_supported("test")
+
+        if isinstance(image_size, int):
+            image_size = (image_size, image_size)
+        super().__init__(
+            train_transform=train_transform,
+            val_transform=val_transform,
+            test_transform=test_transform,
+            predict_transform=predict_transform,
+            image_size=image_size,
+        )
+
+    @property
+    @_apply_to_input
+    def default_train_transforms(self) -> Dict[str, Callable]:
```

Comment:
@ethanwharris @pmeier We made a new addition. This code can be simplified.

Now, {}_default_transforms is a Preprocess hook.

You can also do:
```
def default_transforms(self):
    if not self.training or self.predicting: return None 
    if self.training:
        dict(
            to_tensor_transform=torchvision.transforms.ToTensor(),
            per_sample_transform_on_device=nn.Sequential(
                transforms.Resize(min(self.image_size)),
                transforms.CenterCrop(self.image_size),
            ),
        )
    else:
         dict(
            to_tensor_transform=torchvision.transforms.ToTensor(),
            per_sample_transform_on_device=transforms.Resize(min(self.image_size)),
        )

```

Reply:
I've encountered a problem with this: since I'm inheriting `ImageClassificationPreprocess` the `default_transforms_train` takes a higher priority. I think this is not intended. It happens because "prefix" (should be postfix BTW) is an empty string

https://github.com/PyTorchLightning/lightning-flash/blob/8eef52fa9b72befab339e388ce862e09f5f28e89/flash/core/data/data_pipeline.py#L183

but the check if for overridden methods checks for `None` explicitly:

https://github.com/PyTorchLightning/lightning-flash/blob/8eef52fa9b72befab339e388ce862e09f5f28e89/flash/core/data/data_pipeline.py#L131

Thus, `current_method_name` becomes `_default_transforms` which is indeed not overridden.

Reply:
Fixed in 21d0817.